From 2322c64c2d2fcf68b36dfbd8cb98a347200848d6 Mon Sep 17 00:00:00 2001
From: Dmitry Kasatkin <d.kasatkin@samsung.com>
Date: Fri, 13 Jun 2014 16:31:46 +0300
Subject: [PATCH 07/45] evm: provide a function to set EVM key from the kernel

Crypto HW kernel module can possibly initialize EVM key from the
kernel __init code to enable EVM before calling 'init' process.
This patch provide a function evm_set_key() which can be used to
set custom key directly to EVM without using KEY subsystem.

Signed-off-by: Dmitry Kasatkin <d.kasatkin@samsung.com>
---
 include/linux/evm.h                 |  7 +++++++
 security/integrity/evm/evm_crypto.c | 12 ++++++++++++
 2 files changed, 19 insertions(+)

diff --git a/include/linux/evm.h b/include/linux/evm.h
index 1fcb88c..344c1c4 100644
--- a/include/linux/evm.h
+++ b/include/linux/evm.h
@@ -14,6 +14,7 @@
 struct integrity_iint_cache;
 
 #ifdef CONFIG_EVM
+extern int evm_set_key(void *key, int keylen);
 extern enum integrity_status evm_verifyxattr(struct dentry *dentry,
 					     const char *xattr_name,
 					     void *xattr_value,
@@ -42,6 +43,12 @@ static inline int posix_xattr_acl(const char *xattrname)
 }
 #endif
 #else
+
+static inline int evm_set_key(void *key, int keylen)
+{
+	return -EOPNOTSUPP;
+}
+
 #ifdef CONFIG_INTEGRITY
 static inline enum integrity_status evm_verifyxattr(struct dentry *dentry,
 						    const char *xattr_name,
diff --git a/security/integrity/evm/evm_crypto.c b/security/integrity/evm/evm_crypto.c
index 6e9a9b7..860519e 100644
--- a/security/integrity/evm/evm_crypto.c
+++ b/security/integrity/evm/evm_crypto.c
@@ -32,6 +32,18 @@ struct crypto_shash *hash_tfm;
 
 static DEFINE_MUTEX(mutex);
 
+int evm_set_key(void *key, int keylen)
+{
+	if (evm_initialized & EVM_STATE_KEY_SET)
+		return -EBUSY;
+	if (keylen > MAX_KEY_SIZE)
+		return -EINVAL;
+	memcpy(evmkey, key, keylen);
+	evm_initialized |= EVM_STATE_KEY_SET;
+	return 0;
+}
+EXPORT_SYMBOL_GPL(evm_set_key);
+
 static struct shash_desc *init_desc(char type)
 {
 	long rc;
-- 
2.1.4

